File: sample-code/examples/python/ios_complex.py
"""
More involved iOS tests, using UICatalog application.
"""
import unittest
import os
import random
import string
from appium import webdriver
from appium.webdriver.common.touch_action import TouchAction
# from selenium.webdriver.common.action_chains import ActionChains
from selenium.webdriver.common.touch_actions import TouchActions
# from selenium.webdriver.common.keys import Keys
import urllib2
import json
from time import sleep

def str_generator(size=6, chars=string.ascii_uppercase + string.digits):
    return ''.join(random.choice(chars) for x in range(size))


class ComplexIOSTests(unittest.TestCase):

    def setUp(self):
        # set up appium
        app = os.path.join(os.path.dirname(__file__),
                           '../../apps/UICatalog/build/Release-iphonesimulator',
                           'UICatalog.app')
        app = os.path.abspath(app)
        self.driver = webdriver.Remote(
            command_executor='http://127.0.0.1:4723/wd/hub',
            desired_capabilities={
                'app': app,
                'platformName': 'iOS',
                'platformVersion': '7.1',
                'deviceName': 'iPhone Simulator'
            })
        self._values = []

    def tearDown(self):
        self.driver.quit()

    def _open_menu_position(self, index):
        # Opens up the menu at position [index] : starts at 0.
        table = self.driver.find_element_by_class_name("UIATableView")
        self._row = table.find_elements_by_class_name("UIATableCell")[index]
        self._row.click()

    def test_find_element(self):
        # first view in UICatalog is a table
        table = self.driver.find_element_by_class_name("UIATableView")
        self.assertIsNotNone(table)

        # is number of cells/rows inside table correct
        rows = table.find_elements_by_class_name("UIATableCell")
        self.assertEqual(18, len(rows))

        # is first one about buttons
        self.assertEqual(rows[0].get_attribute("name"), "Action Sheets, AAPLActionSheetViewController")

        # there is nav bar inside the app
        nav_bar = self.driver.find_element_by_class_name("UIANavigationBar")
        self.assertTrue(nav_bar)

    def test_frames(self):
        # go into webview frame
        self._open_menu_position(15)

        # get the contexts and switch to webview
        contexts = self.driver.contexts
        self.assertEqual([u'NATIVE_APP', u'WEBVIEW_1'], contexts)
        self.driver.switch_to.context(contexts[1])

        # Find the store link
        sleep(4) # let the page load, perhaps
        logo = self.driver.find_element_by_id('gn-apple')
        self.assertIsNotNone(logo)

        # leave the webview
        self.driver.switch_to.context(contexts[0])

        # Verify we are out of the webview
        scroll_after = self.driver.find_element_by_class_name("UIAScrollView")
        self.assertTrue(scroll_after)

    def test_location(self):
        # get third row location
        row = self.driver.find_elements_by_class_name("UIATableCell")[2]
        self.assertEqual(row.location['x'], 0)
        self.assertEqual(row.location['y'], 152)

    def test_screenshot(self):
        # make screenshot and get is as base64
        screenshot = self.driver.get_screenshot_as_base64()
        self.assertTrue(screenshot)

        # make screenshot and save it to the local filesystem
        success = self.driver.get_screenshot_as_file("foo.png")
        self.assertTrue(success)
        self.assertTrue(os.path.isfile("foo.png"))

        # get rid of the file
        os.remove("foo.png")

    def test_text_field_edit(self):
        # go to the text fields section
        self._open_menu_position(13)

        text_field = self.driver.find_elements_by_class_name("UIATextField")[0]

        # get default/empty text
        default_val = text_field.get_attribute("value")

        # write some random text to element
        rnd_string = str_generator()
        text_field.send_keys(rnd_string)
        self.assertEqual(text_field.get_attribute("value"), rnd_string)

        # clear and check if is empty/has default text
        text_field.clear()
        self.assertEqual(text_field.get_attribute("value"), default_val)

    def test_alert_interaction(self):
        # go to the alerts section
        self.driver.find_element_by_name('Alert Views, AAPLAlertViewController').click()
        triggerOk = self.driver.find_element_by_accessibility_id("Simple")

        # TOFIX: Looks like alert object is not proper state
        # something to do with UIActionSheet vs. UIAlertView?
        # triggerOk = elements[0]
        triggerOk.click()
        alert = self.driver.switch_to_alert()

        # dismiss alert
        alert.accept()

        # trigger modal alert with cancel & ok buttons
        triggerOkCancel = self.driver.find_element_by_accessibility_id("Okay / Cancel")
        triggerOkCancel.click()
        alert = self.driver.switch_to_alert()

        # check if title of alert is correct
        self.assertEqual(alert.text, "A Short Title Is Best A message should be a short, complete sentence.")
        alert.accept()

    def test_slider(self):
        # go to controls
        self._open_menu_position(10)

        # get the slider
        slider = self.driver.find_element_by_class_name("UIASlider")
        self.assertEqual(slider.get_attribute("value"), "42%")

        slider.set_value(0)
        self.assertEqual(slider.get_attribute("value"), "0%")

    def test_sessions(self):
        data = json.loads(urllib2.urlopen("http://localhost:4723/wd/hub/sessions").read())
        self.assertEqual(self.driver.session_id, data['sessionId'])

    def test_size(self):
        table = self.driver.find_element_by_class_name("UIATableView").size
        row = self.driver.find_elements_by_class_name("UIATableCell")[0].size
        self.assertEqual(table['width'], row['width'])
        self.assertNotEqual(table['height'], row['height'])

    def test_source(self):
        # get main view soruce
        source_main = self.driver.page_source
        self.assertIn("UIATableView", source_main)
        self.assertIn("Text Fields, AAPLTextFieldViewController", source_main)

        # got to text fields section
        self._open_menu_position(13)
        sleep(10)
        source_textfields = self.driver.page_source
        self.assertIn("UIAStaticText", source_textfields)
        self.assertIn("Text Fields", source_textfields)

        self.assertNotEqual(source_main, source_textfields)


suite = unittest.TestLoader().loadTestsFromTestCase(ComplexIOSTests)
unittest.TextTestRunner(verbosity=2).run(suite)

File: sample-code/examples/node/ios-complex.js
"use strict";

require("./helpers/setup");

var wd = require("wd"),
    _ = require('underscore'),
    Q = require('q'),
    serverConfigs = require('./helpers/appium-servers'),
    _p = require('./helpers/promise-utils'),
    fs = require('fs');

describe("ios simple", function () {
  this.timeout(300000);
  var driver;
  var allPassed = true;

  before(function () {
    var serverConfig = process.env.SAUCE ?
      serverConfigs.sauce : serverConfigs.local;
    driver = wd.promiseChainRemote(serverConfig);
    require("./helpers/logging").configure(driver);

    var desired = _.clone(require("./helpers/caps").ios71);
    desired.app = require("./helpers/apps").iosUICatalogApp;
    if (process.env.SAUCE) {
      desired.name = 'ios - complex';
      desired.tags = ['sample'];
    }
    return driver.init(desired);
  });

  after(function () {
    return driver
      .quit()
      .finally(function () {
        if (process.env.SAUCE) {
          return driver.sauceJobStatus(allPassed);
        }
      });
  });

  afterEach(function () {
    allPassed = allPassed && this.currentTest.state === 'passed';
  });

  function clickMenuItem(name) {
    return driver
      .elementByName(name)
      .catch(function () {
        return driver
          .elementByClassName('UIATableView')
          .elementsByClassName('>','UIATableCell')
          .then(_p.filterWithName(name)).first();
      }).click();
  }

  it("should print every menu item", function () {
    return driver
      .elementByClassName('UIATableView')
      .elementsByClassName('>','UIATableCell')
      .then(_p.printNames);
  });

  it("should find an element", function () {
    return driver
      // first view in UICatalog is a table
      .elementByClassName('UIATableView')
        .should.eventually.exist
      // check the number of cells/rows inside the  table
      .elementsByClassName('UIATableCell')
        .then(_p.filterDisplayed)
      .then(function (els) {
        els.should.have.length.above(6);
        return els;
      })
      // various checks
      .first().getAttribute('name')
        .should.become('Action Sheets, AAPLActionSheetViewController')
      .waitForElementByClassName('UIANavigationBar')
        .should.eventually.exist;
  });

  it("should switch context", function () {
    return clickMenuItem('Web View, AAPLWebViewController')
      // get the contexts and switch to webview
      .contexts().should.eventually.deep.equal(
        ['NATIVE_APP','WEBVIEW_1']
      ).context('WEBVIEW_1')
      // find the store link
      .sleep(1000)
      .waitForElementById('gn-apple')
        .should.eventually.exist
      // leave the webview
      .context('NATIVE_APP').sleep(1000)
      //Verify we are out of the webview
      .waitForElementByClassName('UIAScrollView')
        .should.eventually.exist
      // back to main menu
      .back();
  });

  it("should get an element location", function () {
    return driver.elementsByClassName("UIATableCell")
      .then(_p.filterDisplayed)
      .at(2)
      .getLocation()
      .then(function (loc) {
        loc.x.should.equal(0);
        loc.y.should.be.above(100);
      });
  });

  it("should take screenshots", function () {
    return driver
      // base64 screeshot
      .takeScreenshot()
        .should.eventually.exist
      // save screenshot to local file
      .then(function () {
        try {
          fs.unlinkSync('/tmp/foo.png');
        } catch (ign) {}
        fs.existsSync('/tmp/foo.png').should.not.be.ok;
      })
      .saveScreenshot('/tmp/foo.png')
      .then(function () {
        fs.existsSync('/tmp/foo.png').should.be.ok;
      });

  });

  it("should edit a text field", function () {
    var el, defaultValue;
    return clickMenuItem('Text Fields, AAPLTextFieldViewController')
      // get the field and the default/empty text
      .elementByClassName('UIATextField')
        .then(function (_el) {
          el = _el;
          return el.getValue(); })
      .then(function (val) { defaultValue = val; })
      // type something
      .then(function () {
          return el
            .sendKeys('1234 appium')
            .getValue().should.become('1234 appium')
            .elementByName('Done').click().sleep(1000); // dismissing keyboard
      })
      // clear the field
      .then(function () { return el.clear(); })
      .then(function () { el.getValue().should.become(defaultValue); })
      // back to main menu
      .back();
  });

  it("should trigger/accept/dismiss an alert", function () {
    return clickMenuItem('Alert Views, AAPLAlertViewController')
      // trigger simple alert
      .elementByName('Simple').click()
      .alertText().should.eventually.include('A Short Title Is Best')
      .dismissAlert()
      // trigger modal alert with cancel & ok buttons
      .elementByName('Okay / Cancel').click()
      .alertText().should.eventually.include('A Short Title Is Best')
      .acceptAlert()
      // back to main menu
      .back();
  });

  it("should set a slider value", function () {
    var slider;
    return clickMenuItem('Sliders, AAPLSliderViewController')
      // retrieve slider, check initial value
      .elementByClassName("UIASlider")
      .then(function (_slider) { slider = _slider; })
      .then(function () {
        return slider.getValue().should.become('42%');
      })
      // change value
      .then(function () { return slider.setImmediateValue("0%"); })
      .then(function () {
        return slider.getValue().should.become('0%');
      })
      // back to main menu
      .back();
  });

  if (!process.env.SAUCE) {
    it("should retrieve the session list", function () {
      driver.sessions()
      .then(function (sessions) {
        JSON.stringify(sessions).should.include(driver.getSessionId());
      });
    });
  }

  it("should retrieve an element size", function () {
    return Q.all([
      driver.elementByClassName('UIATableView').getSize(),
      driver.elementByClassName('UIATableCell').getSize(),
    ]).then(function (sizes) {
      sizes[0].width.should.equal(sizes[1].width);
      sizes[0].height.should.not.equal(sizes[1].height);
    });
  });

  it("should get the source", function () {
    var mainMenuSource;
    // main menu source
    return driver
      .source().then(function (source) {
        mainMenuSource = source;
        mainMenuSource.should.include('UIAStaticText');
        mainMenuSource.should.include('Text Fields');
      })
      // text fields section source
      .then(function () {
        return clickMenuItem("Text Fields, AAPLTextFieldViewController");
      }).source(function (textFieldSectionSource) {
        textFieldSectionSource.should.include('UIAStaticText');
        textFieldSectionSource.should.include('Text Fields');
        textFieldSectionSource.should.not.equal(textFieldSectionSource);
      })
      // back to main menu
      .back();
  });
});
